File: python/22.11/3.py
from collections import deque

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic, ans, arr = dict(), 0, deque([])

        for ch in s:
            while dic.get(ch):
                dic[arr.popleft()] = False
            
            arr.append(ch)
            ans = max(len(arr), ans)
            dic[ch] = True
        
        return ans

File: typescript/23.08/3.ts
function lengthOfLongestSubstring(s: string): number {
    const visit = {}
    let [i, j, ans] = [0, 0, 0]
    while (i < s.length) {
        if (visit[s[i]] === undefined) {
            visit[s[i]] = 0
        }
        if (visit[s[i]] > 0) {
            visit[s[j++]]--
            ans = Math.max(ans, i - j + 1)
        } else {
            visit[s[i++]]++
        }
    }

    return Math.max(ans, i - j)
};
